File: src/main/java/net/theevilreaper/vulpes/backend/exception/ExceptionHandlerAdvice.java
package net.theevilreaper.vulpes.backend.exception;

import io.micronaut.context.annotation.Requires;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.annotation.Produces;
import io.micronaut.http.server.exceptions.ExceptionHandler;
import jakarta.inject.Singleton;
import net.theevilreaper.vulpes.backend.domain.error.ErrorResponse;

/**
 * Handles each incoming exception and returns a response which contains a better overview of the exception.
 * @author theEvilReaper
 * @version 1.0.0
 * @since 1.0.0
 */
@Produces
@Singleton
public class ExceptionHandlerAdvice implements ExceptionHandler<Throwable, HttpResponse<ErrorResponse>> {

    @Override
    public HttpResponse<ErrorResponse> handle(HttpRequest request, Throwable exception) {
        return HttpResponse.notFound().body(new ErrorResponse.ErrorResponseDTO(exception.getMessage()));
    }
}

File: src/main/kotlin/net/theevilreaper/vulpes/backend/exception/ExceptionHandlerAdvice.kt
package net.theevilreaper.vulpes.backend.exception

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

/**
 * Handles each incoming exception and returns a [ResponseEntity] which contains a better overview over the exception.
 * @author theEvilReaper
 * @version 1.0.0
 * @since 1.0.0
 */
@ControllerAdvice
class ExceptionHandlerAdvice {

    /**
     * Handles each incoming [ResourceNotFoundException] and returns a [ResponseEntity] which contains a better overview over the exception.
     * The information about the thrown exception is stored in the [ErrorResponse] object.
     * @param exception the thrown exception
     */
    @ExceptionHandler(ResourceNotFoundException::class)
    fun handleResourceNotFoundException(exception: ResourceNotFoundException): ResponseEntity<ErrorResponse> {
        val message = "The resource with the id ${exception.message} was not found"
        val error = ErrorResponse(exception.httpMethod.name(), HttpStatus.NOT_FOUND.value(), message)
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error)
    }
}
